website/blog/2018-07-19-whats-happening-with-the-pipeline-proposal.md

```diff
@@ -0,0 +1,168 @@
+
+```js
+promise
+  |> await #
+  |> doubleSay(#, ', ')
+  |> # || throw new TypeError()
+  |> capitalize
+  |> # + '!'
+  |> new User.Message(#)
+  |> await stream.write(#)
+  |> console.log;
+```
+
+Smart Pipelines have a few rules for the placeholder. If a bare identifier is provided to a step in the pipeline, no token is necessary, called ["bare style"](https://github.com/js-choi/proposal-smart-pipelines/blob/master/readme.md#bare-style):
+
+```js
+x |> a;
+x |> f.b;
+```
+
+Unlike Hack, unary functions don't require a placeholder token.
+
+For other expressions, a placeholder (called a "lexical topic token") is required, and the code will throw an early SyntaxError if it is not included in ["topic style"](https://github.com/js-choi/proposal-smart-pipelines/blob/master/readme.md#topic-style):
+
+```js
+10 |> # + 1;
+promise |> await #;
+```
+
+If there are any operators, parentheses (including for method calls), brackets, or anything other than identifiers and dot punctuators, then a topic token is necessary. This avoids footguns and eliminate ambiguity when not using a topic token.
+
+Smart pipelines thus resolve the issue of async in an integrative way, allowing all possible expressions to be emedded in a pipeline; not only `await`, but also `typeof`, `yield`, and another other operator desired.
+
+## Where Babel Comes In
+
+Once all three proposals were fleshed out, we realized discussion and debate were unlikely to resolve the inherent tension between them. We decided the best way to make a decision would be with the feedback of developers, using the proposals in real code. Given Babel's role in the community, we decided to introduce all three of the above proposals into the pipeline operator plugin.
+
+Because these proposals parse slightly differently, support needs to be added to `@babel/parser` (formerly `babylon`), and the parser needs to be configured based on which proposal is its current parse target. The pipeline operator plugin itself thus needs the `"proposal"` option, both to configure babylon as well as its own transformation.
+
+We were working under a compressed timeline for this, as we needed to introduce any breaking changes to babel, `@babel/parser`, as well as the pipeline proposal plugin before babel@7 left beta. We also wanted the plugin to be able to eventually default to whichever of the proposals get accepted, so the configuration option becomes obsolete.
+
+Given these two constraints, we opted to introduce this new configuration option and make it required, forcing users to decide which proposal they want to use in their project. Once a particular proposal gets accepted as the canonical behavior of the pipeline operator, we'll deprecate the `"proposal"` option and make the default whichever gets accepted, and the rejected proposals will be removed in the following major version.
```

Comment:
Should we show the example error message here? Probably not necessary but wondering

Reply:
I don't think that's necessary. The error message should stand on its own.